tests/integration.rs

```diff
@@ -34,6 +35,7 @@ enum DirectoryExtension {
 enum FileExtension {
     Bz,
     Bz2,
+    Bz3,
```

Comment:
Alright, CI problems are blocker number 4 (I'm trying to solve them).

But I just arrived at blocker number 5: tests related to this line of code are failing.

`cargo test --release`:
```
ERROR: corrupt xz stream # file.zip.xz.bz3
ERROR: incomplete frame  # file.sz.zst.bz3
```

Reply:
Is there anything I can help with?

Reply:
yes, we need to figure out why `cargo test` is failing locally.

I believe this is an error in the `bzip3-rs` crate or `bzip3` itself, but we need enough info (or a minimal reproduction example) to open an issue in that repo.

A good point to start debugging is the `cargo test` output.

Reply:
Hi. I found this line fails in test `single_file`:
https://github.com/ouch-org/ouch/blob/6c2e1b6d820f4d898494e609555f0498592fd677/tests/integration.rs#L121

I took the `archive` file and did something like: `cat "/tmp/.tmpWMpRmq/file.sz.sz.zst.bz3.gz.bz2.bz.xz" | xz -d | bzip2 -d | bzip2 -d | gzip -d | bzip3 -d | zstd -d`, with no error occurred. But with `ouch` it says `ERROR: incomplete frame`; don't know why.

The Snappy stream (in this file it's `file.sz.sz`) is empty. Is empty stream considered valid for a Snappy format? Seems Snappy doesn't have an official CLI tool and I can't do decompression in a terminal.

BTW I noticed:
> Extension {
>         compression_formats: [
>             Lzma,
>         ],
>         display_text: "xz",
>     },

Actually `xz` has some difference with `LZMA` format (`xz --format=xz` vs `xz --format=lzma`). IMO `xz` is a container format of `LZMA2` format, which is a successor of the lagacy `LZMA` format.

Reply:
Oh, wait!

Your fix to https://github.com/bczhc/bzip3-rs/pull/8 actually solved the `file.zip.xz.bz3` problem.

Now it's just the `.sz` one, I'm trying to debug it.

Reply:
I ran `cargo test` with the latest commit on my machine and it runs fine, all tests pass. Is the issue only on CI?

Reply:
It's failing locally for me, can you run `git log` and tell me at what commit you're at?

Reply:
I downloaded a zip of `65d1229b6c13aeef9a4c16d73df1212fc8cbfc66` (freijon repo) which tested fine. But when I pull and rebase my repo it fails.

Reply:
OK guys. I also did a test and couldn't see any test failure now on my machine. Both at freijon's [`65d1229b6c13aeef9a4c16d73df1212fc8cbfc66`](https://github.com/freijon/ouch/commit/65d1229b6c13aeef9a4c16d73df1212fc8cbfc66) and `ouch-org:main` applied with freijon's changes (Console log: https://gist.github.com/bczhc/e4b356a2f05cae370e016c1fc47bcf73).

Can anyone tell me what's the issue we're now meeting?

Reply:
Sorry ignore what I said just now. After running `rustup update stable`, I can't reproduce the success case. And also sadly I forgot to check the rust toolchain version before `rustup update stable` so I don't know the toolchain version with which I just pasted the test. Searching in my shell history, the last time I ran `rustup update stable` is 2023-06-13, maybe the previous toolchain version is `stable-2023-06-01`. But with that version, it also fails in `single_file`. I don't know why, and how the hell I just passed those tests (as shown in the console log).

Reply:
I reproduced it. This is an unstable bug. Check it out: https://gist.github.com/bczhc/54a0ffbacd1de7591df40aaf92fc3287. Try searching `single_file ... ok` and `single_file ... failed`.

You can see the first two runs passed, but further runs failed.

Toolchain version: `stable-2023-06-01`

Reply:
So is the root cause coming from [bczhc/bzip3-rs](https://github.com/bczhc/bzip3-rs) then?

Reply:
Yea very likely, my bad... The error is still from chained stuff and I'm still investing it. I tracked this in https://github.com/bczhc/bzip3-rs/issues/9.

Reply:
@freijon Hi. I've located the root cause and fixed it.

One interesting case:
```console
~/t/2 ❯ touch file                                                                                                           17:54:27
~/t/2 ❯ ouch c file file.sz.xz.bz3                                                                                           17:54:29
[INFO] Successfully compressed 'file.sz.xz.bz3'.
~/t/2 ❯ ouch d file.sz.xz.bz3 --yes && cat file                                                                              17:54:31
[INFO] Failed to confirm the format of `file` by sniffing the contents, file might be misnamed
[ERROR] premature eof
~/t/2 ❯                                                                                                                      17:54:35
~/t/2 ❯ rm *                                                                                                                 17:54:35
zsh: sure you want to delete all 2 files in /home/bczhc/t/2 [yn]? y
~/t/2 ❯ printf a > file                                                                                                      17:54:37
~/t/2 ❯ ouch c file file.sz.xz.bz3                                                                                           17:54:41
[INFO] Successfully compressed 'file.sz.xz.bz3'.
~/t/2 ❯ ouch d file.sz.xz.bz3 --yes && cat file                                                                              17:54:42
[INFO] Failed to confirm the format of `file` by sniffing the contents, file might be misnamed
[INFO] Successfully decompressed archive in current directory.
[INFO] Files unpacked: 1
a%
```

Only when the source file is empty the error encounters. I guess, when the source is empty, `XzEncoder` will call `write` once on `Bz3Encoder` with an empty buffer, so that `Bz3Encoder` produces an empty block. `Bz3Decoder` didn't handle empty blocks well, and finally leads to such a bug.

I now suppose empty blocks are totally valid in bzip3 format, and handle them. Another possible solution is reject `write` calls with zero-length buffer so that empty blocks won't be made.

Reply:
Thanks for your help @bczhc :heart:, with all tests passing locally here, I'll mark this thread as resolved.